src/ir/daphneir/DaphneInferTypesOpInterface.cpp

```diff
@@ -276,6 +276,57 @@ std::vector<Type> daphne::OrderOp::inferTypes() {
     return {t};
 }
 
+
+mlir::Type mlirTypeForCode(ValueTypeCode type, Builder builder) {
+    switch(type) {
+        case ValueTypeCode::SI8:  return builder.getIntegerType(8, true);
+        case ValueTypeCode::SI32: return builder.getIntegerType(32, true);
+        case ValueTypeCode::SI64: return builder.getIntegerType(64, true);
+        case ValueTypeCode::UI8:  return builder.getIntegerType(8, false);
+        case ValueTypeCode::UI32: return builder.getIntegerType(32, false);
+        case ValueTypeCode::UI64: return builder.getIntegerType(64, false);
+        case ValueTypeCode::F32: return builder.getF32Type();
+        case ValueTypeCode::F64: return builder.getF64Type();
+        default: throw std::runtime_error("mlirTypeForCode: unknown value type code");
+    }
+}
+
+std::vector<Type> daphne::ReadOp::inferTypes() {
+    FileMetaData fmd = CompilerUtils::getFileMetaData(getFileName());
+    Builder builder(getContext());
+    if (auto resType = getRes().getType().dyn_cast<daphne::MatrixType>()) {
+        mlir::Type valType = mlirTypeForCode(fmd.schema[0], builder);
+        return {mlir::daphne::MatrixType::get(getContext(), valType)};
+    }
+
+    if (auto resType = getRes().getType().dyn_cast<daphne::FrameType>()) {
+        // If an individual value type was specified per column
+        // (fmd.isSingleValueType == false), then this silently uses the
+        // type of the first column.
```

Comment:
This comment about isSingleValueType is not correct and can be removed.

Reply:
Just moved it over, but I can also remove it alltogether.

Reply:
The comment referred to reading a matrix from a file that has a schema defined. So it is not utterly wrong. You could move it a few lines up to the code that handles reading matrices.

Reply:
done.